src/compiler/program.ts

```diff
@@ -626,6 +631,17 @@ namespace ts {
 
         Debug.assert(!!missingFilePaths);
 
+        // List of collected files is complete; validate exhautiveness if this is a project with a file list
+        if (options.composite && rootNames.length < files.length) {

```

Comment:
You might be able to relax/remove this check by using `preProcessFile` from the services layer on the importing side to get the files a given file depends on (recursively) without doing a full parse, for the most part. Still definitely a good practice to explicitly list all your files, but that could reduce friction in adopting the composite flag.

Reply:
When I did the dogfooding conversion, I just used `--listFiles` to fix the files list manually - seemed to be good enough.

Reply:
Any projects that depend on node modules for declarations and/or have a few hundred files I would think may not enjoy managing a list themselves - especially as the exact files in their modules shift under then.

Reply:
If you have a few hundred files, then your tsconfig would almost certainly be globbing. I'd look askance at anyone *not* globbing in that situation - detecting an orphaned .ts file that didn't get imported by anyone would be a big pit of failure.